ReceiveActor - Removing MatchBuilder and making use of handlers.

mhbuck opened the pull request:
Fixes # 
This is part of the changed needed to resolve https://github.com/akkadotnet/akka.net/issues/2811

## Changes
Here the changes are to not make use of the match builder and just make use of the handlers. The approach I have taken is very similar in structure to what was existing but making use of handlers for the various types. There could be merit in removing more of the logic to simplify the class more.

This approach has some internal classes (I created as internal as an initial approach I can see merit in pulling these out to be more clean within the class.

While all of these tests do pass I do think there are some edge cases that this does not cover when reading the comments in the MatchBuilder class. If the general approach of the handlers is what is wanted I will go into the comments and edge cases in more detail.

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).


### Latest `dev` Benchmarks 
Note I pushed the dotnet up to 8.0 because I did not have 6.0 installed on this machine

```
BenchmarkDotNet v0.13.12, Windows 10 (10.0.19045.5371/22H2/2022Update)
Intel Core i7-6700K CPU 4.00GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 9.0.102
  [Host]     : .NET 8.0.12 (8.0.1224.60305), X64 RyuJIT AVX2
  Job-WBRIAI : .NET 8.0.12 (8.0.1224.60305), X64 RyuJIT AVX2

InvocationCount=1  LaunchCount=10  RunStrategy=Monitoring
UnrollFactor=1  WarmupCount=10

| Method                                | Mean     | Error   | StdDev  | Req/sec      |
|-------------------------------------- |---------:|--------:|--------:|-------------:|
| Actor_ping_pong_single_pair_in_memory | 227.4 ns | 1.37 ns | 4.04 ns | 4 397 699,09 |

Warming up...
OSVersion:              Microsoft Windows NT 10.0.19045.0
ProcessorCount:         8
ClockSpeed:             0 MHZ
Actor Count:            16
Messages sent/received: 30000000  (3e7)
Is Server GC:           True
Thread count:           26

ActorBase    first start time: 10.46 ms
ReceiveActor first start time: 16.52 ms

            ActorBase                          ReceiveActor
Throughput, Msgs/sec, Start [ms], Total [ms],  Msgs/sec, Start [ms], Total [ms]
         1, 25210000,      14.80,    1205.11,  28409000,      88.74,    1144.76
         5, 40595000,       7.58,     747.50,  39370000,       2.97,     765.10
        10, 46583000,     165.06,     809.82,  46511000,       2.32,     647.37
        15, 57142000,       2.29,     528.20,  54347000,       2.41,     555.26
        20, 52264000,     100.35,     674.69,  47021000,       3.83,     642.75
        30, 51107000,       1.41,     588.55,  49180000,      40.18,     651.06
        40, 59055000,       1.35,     509.36,  47846000,       1.41,     628.67
        50, 50847000,       1.92,     592.04,  58593000,       2.31,     514.41
        60, 55970000,       2.12,     539.09,  52447000,      19.28,     591.90
        70, 58939000,       1.37,     510.65,  54151000,       1.39,     555.68
        80, 57915000,       1.41,     519.77,  56179000,      12.66,     547.47
        90, 57251000,       1.37,     526.20,  48231000,       1.37,     623.77
       100, 64239000,       1.26,     468.96,  60000000,       1.39,     501.52
       200, 58593000,       1.47,     513.87,  52083000,       1.94,     578.61
       300, 64935000,       1.58,     464.56,  57915000,       1.56,     519.81
       400, 57251000,       4.82,     529.21,  56074000,       1.39,     536.56
       500, 63559000,       1.89,     474.61,  47694000,       1.27,     630.63
       600, 64377000,       1.29,     468.21,  60975000,       1.39,     493.41
       700, 57581000,       1.23,     522.33,  61728000,       1.27,     488.07
       800, 62630000,       1.27,     480.66,  56925000,       1.27,     528.49
       900, 62630000,       1.26,     481.06,  51724000,       1.25,     582.17
```

### This PR's Benchmarks

So as the numbers are showing the performance has gone backwards

```
BenchmarkDotNet v0.13.12, Windows 10 (10.0.19045.5371/22H2/2022Update)
Intel Core i7-6700K CPU 4.00GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 9.0.102
  [Host]     : .NET 8.0.12 (8.0.1224.60305), X64 RyuJIT AVX2
  Job-WLSTQM : .NET 8.0.12 (8.0.1224.60305), X64 RyuJIT AVX2

InvocationCount=1  LaunchCount=10  RunStrategy=Monitoring
UnrollFactor=1  WarmupCount=10

| Method                                | Mean     | Error   | StdDev  | Req/sec      |
|-------------------------------------- |---------:|--------:|--------:|-------------:|
| Actor_ping_pong_single_pair_in_memory | 236.5 ns | 1.77 ns | 5.22 ns | 4 228 340,67 |

Warming up...
OSVersion:              Microsoft Windows NT 10.0.19045.0
ProcessorCount:         8
ClockSpeed:             0 MHZ
Actor Count:            16
Messages sent/received: 30000000  (3e7)
Is Server GC:           True
Thread count:           26

ActorBase    first start time: 11.96 ms
ReceiveActor first start time:  5.37 ms

            ActorBase                          ReceiveActor
Throughput, Msgs/sec, Start [ms], Total [ms],  Msgs/sec, Start [ms], Total [ms]
         1, 22692000,       5.69,    1328.11,  28544000,      44.97,    1096.91
         5, 44247000,     156.34,     834.79,  39421000,       3.50,     765.35
        10, 57692000,       2.62,     523.27,  40983000,      33.38,     765.45
        15, 52356000,      12.17,     585.59,  52631000,     145.33,     716.11
        20, 62240000,       3.25,     486.14,  51282000,       4.10,     589.88
        30, 62893000,       1.72,     479.48,  44247000,       9.63,     688.59
        40, 63965000,       2.06,     471.63,  53003000,       1.58,     568.51
        50, 63424000,      18.76,     492.70,  48154000,       1.57,     625.47
        60, 67567000,       1.38,     445.43,  53571000,       1.56,     561.95
        70, 59405000,       1.42,     506.46,  54945000,       1.40,     548.33
        80, 66371000,       1.38,     454.23,  55045000,       2.75,     548.59
        90, 61983000,       1.28,     485.55,  55248000,       1.53,     544.57
       100, 65502000,       1.35,     460.24,  52356000,       1.49,     574.49
       200, 67873000,       1.93,     444.42,  55970000,       1.49,     537.89
       300, 63694000,       1.63,     472.91,  54347000,      62.60,     614.65
       400, 63965000,       1.74,     471.53,  51194000,       1.36,     587.42
       500, 68807000,       1.39,     438.34,  52447000,       1.46,     573.91
       600, 58479000,       1.29,     514.95,  54844000,       1.60,     548.89
       700, 53956000,       1.30,     557.95,  46296000,       1.85,     650.38
       800, 62111000,       2.65,     486.63,  51282000,       1.61,     587.52
       900, 65075000,       1.69,     463.57,  46224000,       1.95,     651.46
```

mhbuck committed:
Initial working approach but very rough.

mhbuck committed:
Consolidate

mhbuck committed:
Potentially cleaner approach

mhbuck committed:
Some clean up and comments.

mhbuck committed:
Remove using

mhbuck commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -20,10 +19,51 @@ namespace Akka.Actor
+        // Moved out and better tested
+        class HandlerSet
+        {
+            public HandlerSet()
```
The idea here is having something that handles the calls for each of the states of the actor.
1. TypedHandlers - Handlers when we have the type information
2. HandleObject - Handles calls when the type information supplied is object
3. Handle Any - The catch all handler 

Now that I am reviewing this I feel like the naming should be  a bit more consistent

mhbuck commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -20,10 +19,51 @@ namespace Akka.Actor
+        }
+
+        // Attempting to make more use of generics to avoid boxing
+        class TypeHandler<T> : ITypeHandler
```
Have an interface that impliments some methods that can be used during the playback of messages. This was seemingly the nicest way to be able to handle generics and handle object calls.

to11mtm replied:
Pattern makes sense to me. 

That said, I wonder if A more aggressive 'TryHandle(object message)` is warranted (i.e. to rely less on PGO/etc to avoid multiple unboxes) but benchmarks can be used to check benefit if any.

mhbuck commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -312,7 +415,79 @@ protected void Receive(Type messageType, Func<object, bool> handler)
+                handlers.TypedHandlers[typeof(T)] = handlerList;
+            }
+
+            var typeHandler = new TypeHandler<T>
```
Handlers where the Generic types are present and can be used

mhbuck commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -312,7 +415,79 @@ protected void Receive(Type messageType, Func<object, bool> handler)
+
+            if (typeof(T) == typeof(object))
+            {
+                handlers.HandleObject.Add(new TypeHandler<object>
```
If the type supplied is object needed to make use of an object type handler. This is one of the things that I was not a big fan of but could not think of a more elegant way of doing it.

mhbuck commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -312,7 +415,79 @@ protected void Receive(Type messageType, Func<object, bool> handler)
+            }
+
+            // Have to use object here as dont have the generic type information
+            var typeHandler = new TypeHandler<object>
```
Because the generic information is not present here we needed to resort to using the object type.

mhbuck reviewed (commented):
This was my approach that does work but has some work I feel. There are definitely some edge cases that still need to be checked before finalising.

Aaronontheweb commented:
@mhbuck I'm getting 351 test failures here so there might be an issue with the way we're doing message handling under this PR

mhbuck committed:
Added test and logic to prevent Any handler from being added twice

mhbuck commented:
Will expand running the tests locally and narrow it down. 👍

mhbuck commented:
> @mhbuck I'm getting 351 test failures here so there might be an issue with the way we're doing message handling under this PR

I am pretty sure it is down to handling of concrete types vs interfaces etc. I took a pretty naive approach for handling. I will have a look into the current logic and work out a better approach.

to11mtm commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -20,10 +19,51 @@ namespace Akka.Actor
+                HandleObject = new List<ITypeHandler>();
+            }
+
+            public Dictionary<Type, List<ITypeHandler>> TypedHandlers { get; set; }
```
Where this can get _tricky_, is that I am not sure whether the overall implementation has the same 'inheritance/execution rules' as existing MatchBuilder.

A fun test case would be

```
public interface IFoo
{
}

public class FooA : IFoo
{
}
public class FooB : IFoo
{
}
///

Receive<IFoo>(/**/);
Receive<FooA>(/**/);
Receive<FooB>(/**/);
```

possibly also playing with predicate rules.

mhbuck replied:
Yeah this scenario is what is breaking most of the tests in clustering because I believe it makes use of this. I am working on building out the tests to cover the inheritance cases and get a better understanding the current matchbuilder functionality and where my approach is lacking. It will also make the receive actor tests more robust.

These tests are what I need to dig into more and understand better https://github.com/akkadotnet/akka.net/tree/dev/src/core/Akka.Tests/MatchHandler

to11mtm commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -20,10 +19,51 @@ namespace Akka.Actor
+                HandleObject = new List<ITypeHandler>();
+            }
+
+            public Dictionary<Type, List<ITypeHandler>> TypedHandlers { get; set; }
```
But also, any reasons for not having it be a get-only? I get that `HandleAny` can mutate but not sure about this one (or `HandleObject`

to11mtm commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -70,14 +104,56 @@ protected sealed override void OnReceive(object message)
         {
-            var wasHandled = partialAction(message);
-            if(!wasHandled && _shouldUnhandle)
+            var messageType = message.GetType();
```
this should probably get inlined into the `TryGetValue` (or if we need a different lookup call).

to11mtm commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -70,14 +104,56 @@ protected sealed override void OnReceive(object message)
+                currentHandler.HandleAny(message);
+            }
+
+            if (_shouldUnhandle)
```
Might be a stupid question, but should this also get hit if `HandleAny` is null? (or maybe other logic concern?)

to11mtm reviewed (commented):
Left some initial feedback, need to pull it down and look further though.

Impressive at first look!

Edited to add:

Didn't look at benchmark numbers first, Now that I have that gives me some things to dig deeper into. I think some of the feedback may help, It may also be worth looking at profiling to see where hotspots are and how those can be mitigated.

mhbuck committed:
Wanting to confirm these tests are also passing on build

This is probably not the most ideal code. I want to make sure the tests are passing better before getting to cleaning it up

mhbuck committed:
TryHandle looks much neater

mhbuck committed:
Closer to what I think the final layout will be

mhbuck committed:
Moved classes out and starting on tests

Still quite a bit of understanding of how the MatchBuilder was working

mhbuck committed:
Test update.

mhbuck committed:
More refinement

mhbuck committed:
Getting more tests

mhbuck committed:
Covering a bunch of the test cases

mhbuck commented on src/core/Akka/Actor/ReceiveActorHandlers.cs:
```diff
@@ -0,0 +1,163 @@
+
+    private Action<object>? HandleAny { get; set; }
+
+    private void CanAddMoreHandlers()
```
Validation cases if more handlers can be added
1. If an object is handled and there is no predicate it is a catch all so anything after that would not be valid
2. If HandleAny is added

Side Note - Conceptually these two are the same. It has been separated but possibly could be joined.

mhbuck commented on src/core/Akka/Actor/ReceiveActorHandlers.cs:
```diff
@@ -0,0 +1,163 @@
+        }
+    }
+    
+    public void AddGenericReceiveHandler<T>(Predicate<T>? shouldHandlePredicate, Func<T, bool> handler)
```
If the generic type information is present then it can be used.

mhbuck commented on src/core/Akka/Actor/ReceiveActorHandlers.cs:
```diff
@@ -0,0 +1,163 @@
+        var typedHandler = (TypeHandler<object>)typeHandlerInterface;
+
+        // Have to use object here as dont have the generic type information
+        var predicateHandler = new PredicateHandler<object>(shouldHandlePredicate, handler);
```
Using Typed methods need to make use of object type.

mhbuck commented on src/core/Akka/Actor/ReceiveActorHandlers.cs:
```diff
@@ -0,0 +1,163 @@
+        {
+            // This is covering object types as well. There might be an ordering issue here
+            // but this should probably be resolved with the logic around how handlers are ordered.
+            if (type.IsAssignableFrom(messageType))
```
While going through the various handlers if the message type is assignable to the type the handlers can use then try execute.

mhbuck commented on src/core/Akka/Actor/ReceiveActorHandlers.cs:
```diff
@@ -0,0 +1,163 @@
+
+    public bool TryHandle(object message)
+    {
+        var typedMessage = (T)message;
```
Once in the Handling for one specific type then cast the message into that type once and pass it into the try handle method

mhbuck commented on src/core/Akka/Actor/ReceiveActorHandlers.cs:
```diff
@@ -0,0 +1,163 @@
+    public Predicate<T>? Predicate { get; }
+    public Func<T, bool> Handler { get; }
+
+    public bool TryHandle(T typedMessage)
```
If this predicate and func handler combination can handle it execute or return false.

mhbuck commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -312,7 +337,27 @@ protected void Receive(Type messageType, Func<object, bool> handler)
+        
+        // Separated the handling of the generic and typed handlers because the generics it can be handled
+        // one way but using types needed to assume object types.
+        private void AddGenericReceiveHandler<T>(Predicate<T> shouldHandle, Func<T, bool> handler)
```
There is still some investigation that I need to do about turning all of the Actions into Func's and how that is all handled. I feel like there is some wild edge cases that might get tripped up here.

mhbuck commented on src/core/Akka/Actor/ReceiveActorHandlers.cs:
```diff
@@ -0,0 +1,163 @@
+    }
+}
+
+internal interface ITypeHandler
```
Would you like all these classes in separate files? Any preference on the namespace they should be in. 

I am also aware the names are maybe not descriptive enough.

mhbuck reviewed (commented):
This is in a very close state to being ready. I have added the tests to try replicate what is being done here https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka.Tests/MatchHandler/MatchHandlerBuilderTests.cs 

I am still looking into a couple cases around how the Func's are being handled. There might be some interesting edge cases that I am missing

mhbuck committed:
Minor Tweak

mhbuck commented:
Latest Benchmarks

**Before Change**
```
BenchmarkDotNet v0.13.12, Windows 10 (10.0.19045.5487/22H2/2022Update)
Intel Core i7-6700K CPU 4.00GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 9.0.102
  [Host]     : .NET 8.0.12 (8.0.1224.60305), X64 RyuJIT AVX2
  Job-OKKNQX : .NET 8.0.12 (8.0.1224.60305), X64 RyuJIT AVX2

InvocationCount=1  LaunchCount=10  RunStrategy=Monitoring
UnrollFactor=1  WarmupCount=10

| Method                                | Mean     | Error   | StdDev  | Req/sec      |
|-------------------------------------- |---------:|--------:|--------:|-------------:|
| Actor_ping_pong_single_pair_in_memory | 227.9 ns | 1.43 ns | 4.23 ns | 4 387 834,09 |

Ping Pong
OSVersion:              Microsoft Windows NT 10.0.19045.0
ProcessorCount:         8
ClockSpeed:             0 MHZ
Actor Count:            16
Messages sent/received: 30000000  (3e7)
Is Server GC:           True
Thread count:           26

ActorBase    first start time:  7.23 ms
ReceiveActor first start time: 18.99 ms

            ActorBase                          ReceiveActor
Throughput, Msgs/sec, Start [ms], Total [ms],  Msgs/sec, Start [ms], Total [ms]
         1, 24057000,     111.05,    1358.32,  24469000,     115.64,    1341.86
         5, 39062000,      16.47,     784.60,  39893000,       4.22,     756.33
        10, 43604000,       3.39,     692.36,  45454000,       4.28,     664.38
        15, 57034000,       2.77,     528.90,  51993000,       2.32,     579.61
        20, 49586000,      23.89,     629.86,  52356000,       3.45,     576.73
        30, 58365000,       1.58,     515.81,  46801000,       1.37,     642.70
        40, 62111000,       1.34,     485.30,  51546000,       1.45,     584.12
        50, 52910000,       1.47,     568.76,  49180000,       2.03,     612.23
        60, 61099000,       1.56,     493.42,  54744000,       1.66,     550.64
        70, 59055000,       1.42,     509.71,  59055000,       1.39,     509.48
        80, 49586000,       1.56,     606.87,  49423000,       1.37,     608.59
        90, 64102000,       1.43,     470.09,  57581000,       1.39,     522.40
       100, 62893000,       1.39,     479.06,  56285000,       1.70,     535.58
       200, 57803000,       1.36,     520.62,  55452000,       1.68,     543.05
       300, 53475000,       1.76,     563.48,  58593000,       1.61,     513.69
       400, 57915000,       1.84,     520.35,  59880000,       1.64,     503.34
       500, 55658000,       1.33,     540.33,  60606000,       2.40,     498.05
       600, 64239000,      18.50,     485.94,  60120000,       1.37,     501.04
       700, 58252000,       1.55,     517.49,  56925000,       1.57,     528.88
       800, 60606000,       1.35,     497.16,  52910000,       1.54,     568.78
       900, 59642000,       2.12,     505.87,  57471000,       1.68,     523.71
```
**PR Benchmark**
```
BenchmarkDotNet v0.13.12, Windows 10 (10.0.19045.5487/22H2/2022Update)
Intel Core i7-6700K CPU 4.00GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 9.0.102
  [Host]     : .NET 8.0.12 (8.0.1224.60305), X64 RyuJIT AVX2
  Job-QTOHXP : .NET 8.0.12 (8.0.1224.60305), X64 RyuJIT AVX2

InvocationCount=1  LaunchCount=10  RunStrategy=Monitoring
UnrollFactor=1  WarmupCount=10

| Method                                | Mean     | Error   | StdDev  | Req/sec      |
|-------------------------------------- |---------:|--------:|--------:|-------------:|
| Actor_ping_pong_single_pair_in_memory | 240.2 ns | 1.22 ns | 3.59 ns | 4 162 677,83 |



OSVersion:              Microsoft Windows NT 10.0.19045.0
ProcessorCount:         8
ClockSpeed:             0 MHZ
Actor Count:            16
Messages sent/received: 30000000  (3e7)
Is Server GC:           True
Thread count:           26

ActorBase    first start time: 15.90 ms
ReceiveActor first start time: 17.30 ms

            ActorBase                          ReceiveActor
Throughput, Msgs/sec, Start [ms], Total [ms],  Msgs/sec, Start [ms], Total [ms]
         1, 20775000,      35.37,    1480.08,  24732000,     152.29,    1366.12
         5, 44182000,     139.31,     818.86,  36496000,       2.95,     825.61
        10, 49180000,       7.30,     617.84,  40214000,       5.45,     751.57
        15, 50933000,       2.34,     591.53,  44709000,      12.84,     684.64
        20, 56285000,      97.15,     630.63,  43165000,       4.28,     699.96
        30, 56925000,       1.31,     528.57,  45180000,       2.19,     666.69
        40, 56818000,       2.05,     530.58,  47318000,       1.43,     636.42
        50, 66964000,       1.68,     449.80,  53763000,       1.43,     560.37
        60, 54744000,       1.39,     549.80,  55658000,       1.71,     541.37
        70, 68493000,       1.31,     439.34,  45317000,       5.20,     667.51
        80, 65645000,       1.31,     458.73,  54446000,       1.33,     552.70
        90, 68337000,       3.65,     442.69,  56390000,       1.37,     533.67
       100, 66815000,       1.26,     450.27,  52264000,       1.47,     576.35
       200, 68337000,       1.40,     440.62,  57581000,       1.60,     522.96
       300, 70422000,       1.30,     427.42,  56710000,       1.80,     531.06
       400, 58139000,       1.69,     517.74,  57142000,       1.76,     527.69
       500, 69767000,       1.25,     431.72,  55658000,       1.29,     540.82
       600, 70093000,       2.51,     430.69,  56390000,      22.96,     555.07
       700, 61349000,       1.35,     490.61,  55452000,       1.30,     543.24
       800, 69124000,       1.25,     435.32,  52356000,       1.33,     575.12
       900, 70257000,       1.22,     428.49,  57471000,       1.28,     523.76
```

Aaronontheweb commented:
> While all of these tests do pass I do think there are some edge cases that this does not cover when reading the comments in the MatchBuilder class. If the general approach of the handlers is what is wanted I will go into the comments and edge cases in more detail.

We exercise `ReceiveActor` pretty aggressively internally - if there are real-world scenarios in actual use that we don't cover, I think we'll discover those during the betas of Akka.NET v1.6.

Aaronontheweb commented:
PR benchmark numbers actually look like an improvement, which doesn't surprise me - the stuff we created via `Reflection.Emit` doesn't benefit from PGO and other JIT improvements.

Aaronontheweb commented on src/core/Akka.Tests/Actor/ReceiveActorHandlersTests.cs:
```diff
@@ -0,0 +1,140 @@
+public class ReceiveActorHandlersTests
+{
+    [Fact]
+    public void Given_ReceiveAnyHandler_Added_When_Adding_Any_Other_Handler_Then_Should_Fail()
```
LGTM

Aaronontheweb commented on src/core/Akka.Tests/Actor/ReceiveActorHandlersTests.cs:
```diff
@@ -0,0 +1,140 @@
+
+        // As a handler for the type of object with no predicate is added,
+        // adding another handler for the same type combination should fail with an exception
+        Assert.Throws<InvalidOperationException>(() =>
```
LGTM

Aaronontheweb commented on src/core/Akka.Tests/Actor/ReceiveActorHandlersTests.cs:
```diff
@@ -0,0 +1,140 @@
+
+        // The handler added has a predicate which makes it uncertain if it will handle the message.
+        // Adding another handler for the same type combination should be allowed.
+        handlers.AddTypedReceiveHandler(typeof(object), null, _ => true);
```
LGTM

Aaronontheweb commented on src/core/Akka.Tests/Actor/ReceiveActorHandlersTests.cs:
```diff
@@ -0,0 +1,140 @@
+        Assert.Throws<InvalidOperationException>(() =>
+            handlers.AddTypedReceiveHandler(typeof(int), _ => true, _ => true));
+        Assert.Throws<InvalidOperationException>(() =>
+            handlers.AddGenericReceiveHandler<bool>(_ => true, _ => true));
```
LGTM

Aaronontheweb commented on src/core/Akka.Tests/Actor/ReceiveActorHandlersTests.cs:
```diff
@@ -0,0 +1,140 @@
+        handlers.AddTypedReceiveHandler(typeof(object), null, _ => true);
+
+        // This should throw because the object handler is already added and would catch this before.
+        Assert.Throws<InvalidOperationException>(() =>
```
IMHO, not a breaking change - it's closer to a bugfix.

Honestly we should add some analyzer rules for this stuff so we can catch it at compile-time - we could and should do that today with Akka.NET v1.5. cc @Arkatufus

Aaronontheweb commented on src/core/Akka.Tests/Actor/ReceiveActorHandlersTests.cs:
```diff
@@ -0,0 +1,140 @@
+
+        // The handler added has a predicate which makes it uncertain if it will handle the message.
+        // Adding another handler for the same type combination should be allowed.
+        handlers.AddGenericReceiveHandler<int>(null, _ => true);
```
LGTM

Aaronontheweb commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -38,28 +38,22 @@ void IInitializableActor.Init()
         private void PrepareConfigureMessageHandlers()
         {
-            _matchHandlerBuilders.Push(new MatchBuilder(CachedMatchCompiler<object>.Instance));
+            _handlersStack.Push(new ReceiveActorHandlers());
```
LGTM

Aaronontheweb commented on src/core/Akka/Actor/ReceiveActor.cs:
```diff
@@ -133,7 +130,11 @@ private static Action<T> WrapAsyncHandler<T>(Func<T, Task> asyncHandler)
-            Receive(WrapAsyncHandler(handler), shouldHandle);
+            AddGenericReceiveHandler<T>(shouldHandle, message =>
+            {
+                WrapAsyncHandler(handler)(message);
```
LGTM

Aaronontheweb commented on src/core/Akka/Actor/ReceiveActorHandlers.cs:
```diff
@@ -0,0 +1,162 @@
+        HandleAny = null;
+    }
+
+    private Dictionary<Type, ITypeHandler> TypedHandlers { get; }
```
So this might be a bit of an issue: if I iterate through the Dictionary, does that traversal order match the insertion order? Otherwise, we can get unpredictable results if a message might possibly match multiple predicates / type filters inside the `ReceiveActorHandlers`. 

The advice we've been telling users for a decade now is that "`Receive<T>` handlers are always evaluated in the original order that they are declared" - the behavior the `MatchBuilder` used to enforce this was a `List<T>` of `ITypeHandler`s so we could preserve that order. If `Dictionary<TKey, TValue>` doesn't behave that same way that would be a very subtle but immediately noticeable breaking behavioral change for users upgrading.

The fact that all of the tests covering the `ReceiveActorHandlers` pass so far tells me that maybe I'm worried about nothing, but only the paranoid survive.

Aaronontheweb replied:
Yeah it looks like I'm right to be concerned here - the insertion order is not guaranteed to be preserved: https://stackoverflow.com/a/13998263

Aaronontheweb reviewed (changes requested):
So there's a couple of things that merit further investigation:

1. My comment about `Dictionary` traversal ordering
2. Adding some tests where we we have `T:InterfaceOfThing`, a `Receive<InterfaceOfThing>`, and the ability to guarantee that `T` gets handled when sent to it. I am close to 100% certain that given the vastness of the Akka.NET test suite that this is covered somewhere, albeit not directly in your tests.

Aaronontheweb commented:
This PR was included inside of https://github.com/akkadotnet/akka.net/pull/7557 - well done @mhbuck - please feel free to do additional reviews, updates, and changes in a new PR or GH issue.
